engines/wintermute/ad/ad_actor_3dx.cpp

```diff
@@ -0,0 +1,2534 @@
+		} else {
+			getNextStep3D();
+		}
+
+		_modelX->playAnim(0, _walkAnimName, _defaultTransTime, false, _defaultStopTransTime);
+		break;
+
+	//////////////////////////////////////////////////////////////////////////
+	case STATE_TALKING: {
+		_sentence->update();
+
+		if (_sentence->_currentSkelAnim) {
+			_tempSkelAnim = _sentence->_currentSkelAnim;
+		}
+
+		bool timeIsUp = (_sentence->_sound && _sentence->_soundStarted && (!_sentence->_sound->isPlaying() && !_sentence->_sound->isPaused())) ||
+						(!_sentence->_sound && _sentence->_duration <= _gameRef->_currentTime - _sentence->_startTime);
+		if (_tempSkelAnim == nullptr || !_modelX->isAnimPending(0, _tempSkelAnim) || timeIsUp) {
+			if (timeIsUp) {
+				_sentence->finish();
+				_tempSkelAnim = nullptr;
+				_state = _nextState;
+				_nextState = STATE_READY;
+
+				if (_talkAnimChannel > 0)
+					_modelX->stopAnim(_talkAnimChannel, _defaultStopTransTime);
+			} else {
+				_tempSkelAnim = _sentence->getNextStance();
+				if (_tempSkelAnim)
+					_modelX->playAnim(0, _tempSkelAnim, _defaultTransTime, true, _defaultStopTransTime);
+				else {
+					if (_modelX->getAnimationSetByName(_talkAnimName))
+						_modelX->playAnim(_talkAnimChannel, _talkAnimName, _defaultTransTime, false, _defaultStopTransTime);
+					else
+						_modelX->playAnim(0, _idleAnimName, _defaultTransTime, false, _defaultStopTransTime);
+				}
+
+				((AdGame *)_gameRef)->addSentence(_sentence);
+			}
+		} else {
+			if (_tempSkelAnim) {
+				_modelX->playAnim(0, _tempSkelAnim, _defaultTransTime, false, _defaultStopTransTime);
+			}
+
+			((AdGame *)_gameRef)->addSentence(_sentence);
+		}
+		break;
+	}
+
+	case STATE_PLAYING_ANIM:
+		if (_stateAnimChannel != 0) {
+			_modelX->playAnim(0, _idleAnimName, _defaultTransTime, false, _defaultStopTransTime);
+		}
+		break;
+
+	//////////////////////////////////////////////////////////////////////////
+	case STATE_READY:
+		_modelX->playAnim(0, _idleAnimName, _defaultTransTime, false, _defaultStopTransTime);
+		break;
+
+	case STATE_IDLE:
+	case STATE_TURNING_LEFT:
+	case STATE_TURNING_RIGHT:
+	case STATE_PLAYING_ANIM_SET:
+	case STATE_NONE:
+		break;
+	} // switch(_state)
+
+	// finished playing animation?
+	if (_state == STATE_PLAYING_ANIM && !_modelX->isAnimPending(_stateAnimChannel)) {
+		_state = _nextState;
+		_nextState = STATE_READY;
+	}
+
+	updateBlockRegion();
+	_ready = (_state == STATE_READY);
+
+	// setup 2D position
+	int origX = _posX;
+	int origY = _posY;
+
+	bool ret = AdObject3D::update();
+
+	if (origX != _posX || origY != _posY) {
+		afterMove();
+	}
+
+	if (_modelX) {
+		_modelX->update();
+
+		if (_shadowModel) {
+			_shadowModel->update();
+		}
+	}
+
+	updateAttachments();
+	updatePartEmitter();
+
+	return ret;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool AdActor3DX::display() {
+	if (!_modelX) {
+		return true;
+	}
+
+	BaseRenderOpenGL3D *renderer = (BaseRenderOpenGL3D *)_gameRef->_renderer;

```

Comment:
Essentially that's the way that Wintermute3D handles it, I just copied it for the moment but would go with the extra pointer the the gameRef as well.

Reply:
Removed some of the casts, since they were indeed unnecessary. Adding a typed pointer to into the gameRef would a require a cast as well at the moment, since the factory functions only return a BaseRenderer pointer. This cast would be very safe though. Do it like this?

Reply:
Ok, I just added a typed pointer to the gameRef, inserted a cast during the creation of the renderer and otherwise remove all other renderer casts. This should make it easy to add a super class for the 3d renderers later as well.